Enable leak checks

kripken opened the pull request:
Let's see where things stand now.

jfbastien commented on CMakeLists.txt:
```diff
@@ -63,7 +63,8 @@ ELSE()
     ADD_COMPILE_FLAG("-O0")
     ADD_COMPILE_FLAG("-g3")
```
Hoist out this `-g3` instead of duplicating below?

kripken replied:
Oh, this is just throwaway code for testing on travis. Just makes the stack traces from leaks nicer to read. I'll remove when this PR is done and the leaks are gone.

kripken commented:
Ok, this should be good to go. I rebased now and removed the debug help stuff, and all leaks are gone.

A bunch of code changed here, but aside from the `.travis.yaml` change to enable leak checks, it's nothing interesting, just moving to rely on `std::unique_ptr` etc. to manage memory in more places, and removing some hacks.

jfbastien commented on src/asm2wasm.h:
```diff
@@ -216,13 +216,13 @@ class Asm2WasmBuilder {
     assert(ast[0] == CALL && ast[1][0] == NAME);
     IString importName = ast[1][1]->getIString();
-    auto* type = allocator.alloc<FunctionType>();
+    std::unique_ptr<FunctionType> type = make_unique<FunctionType>();
```
`auto`

kripken replied:
fixed.

jfbastien commented on src/emscripten-optimizer/simple_ast.h:
```diff
@@ -84,6 +84,7 @@ struct Arena {
   int arr_index;
 
   Arena() : index(0), arr_index(0) {}
+  ~Arena();
```
?

kripken replied:
This declares a destructor? Not sure what the question is?

jfbastien replied:
Sorry, should have elaborated: would have been nicer to have a container instead, so you can use the default dtor.

jfbastien commented on src/wasm-binary.h:
```diff
@@ -1549,15 +1548,13 @@ class WasmBinaryBuilder {
+      auto offset = getU32LEB();
       auto size = getU32LEB();
-      auto buffer = (char*)malloc(size);
+      char buffer[size];
```
Dynamic alloca :(

kripken replied:
What is the downside of that?

What's a better alternative?

jfbastien replied:
Small stack allocations are fine, but this one is controlled by the input, which is a Bad Idea. Dynamic allocation is better in those cases.

jfbastien commented on src/wasm-s-parser.h:
```diff
@@ -1068,7 +1066,7 @@ class SExpressionWasmBuilder {
       assert(curr[0]->str() == SEGMENT);
       const char *input = curr[2]->c_str();
-      char *data = (char*)malloc(strlen(input)); // over-allocated, since escaping collapses, but whatever
+      char data[strlen(input)];
```
Dynamic alloca :(

kripken commented:
Merging as this touches so much code, it's a merge collision hazard. Let's do additional comments as followups. @jfbastien: I already addressed the first of your comments, and had some questions about the others.

@dschuff: about merge vs squash commits, this pr is an example of where i think a merge commit is better. It's lots of tiny separate commits, not worth their own pr, and it's nice to keep them separate for future bisection and readability purposes. Also the merge commit shows the structure (that those commits led up to the commit enabling leak tests now that they pass).
